openlibrary/core/wikidata.py

```diff
@@ -0,0 +1,123 @@
+"""
+The purpose of this file is to:
+1. Interact with the Wikidata API
+2. Store the results
+3. Make the results easy to access from other files
+"""
+import requests
+import web
+import dataclasses
+from dataclasses import dataclass
+from openlibrary.core.helpers import days_since
+
+from datetime import datetime
+import json
+from openlibrary.core import db
+
+
+@dataclass
+class WikiDataAPIResponse:
+    type: str
+    labels: dict
+    descriptions: dict
+    aliases: dict
+    statements: dict
+    sitelinks: dict
+    id: str
+
+    @classmethod
+    def from_dict(cls, data: dict):
+        # use [''] instead of get so this fails if fields are missing
+        return cls(
+            type=data['type'],
+            labels=data['labels'],
+            descriptions=data['descriptions'],
+            aliases=data['aliases'],
+            statements=data['statements'],
+            sitelinks=data['sitelinks'],
+            id=data['id'],
+        )
+
+
+@dataclass
+class WikiDataEntity:
+    id: str
+    data: WikiDataAPIResponse
+    updated: datetime
+
+    def description(self, language: str = 'en') -> str | None:
+        """If a description isn't available in the requested language default to English"""
+        return self.data.descriptions.get(language) or self.data.descriptions.get('en')
+
+    @classmethod
+    def from_db_query(cls, response: web.utils.Storage):
+        return cls(
+            id=response.id,
+            data=WikiDataAPIResponse.from_dict(response.data),
+            updated=response.updated,
+        )
+
+    @classmethod
+    def from_web(cls, response: dict):
+        data = WikiDataAPIResponse.from_dict(response)
+        return cls(
+            id=data.id,
+            data=data,
+            updated=datetime.now(),
+        )
+
+
+def get_wikidata_entity(QID: str, ttl_days: int = 30) -> WikiDataEntity | None:
```

Comment:
This is only ever set to 30 or 0. We should simplify this to `allow_cache = True` or something. `ttl` is a bit of a technical term which isn't really relevant outside the users of this function. You can put the ttl as a constant at the top of this file ; `WIKIDATA_CACHE_TTL_DAYS = 30`. Or `WIKIDATA_CACHE_TTL = timedelta(days=30)` might be better to use the language native time features!

Reply:
https://en.wikipedia.org/wiki/Time_to_live ;-)

Reply:
Fixed